ISPN-12810 marshalling and encoding title (#9106)
infinispan/infinispan: documentation/src/main/asciidoc/topics/ref_protostream_types.adoc

[id='protostream_types-{context}']
= ProtoStream types

> {brandname} uses a ProtoStream API that can handle the following types for keys and values, as well as the unboxed equivalents in the case of primitive types:

Reviewer: ```suggestion
{brandname} uses the ProtoStream library, which can natively handle the following types for keys and values, as well as the unboxed equivalents in the case of primitive types:
```

I think I've noticed one more place where the text says "ProtoStream API" but it's actually talking about the implementation, not the API.

* `java.util.Date`
* `java.time.Instant`

[discrete]
== Additional type collections

> The ProtoStream library includes an `Adapter` class that you can use to marshall additional types.

Reviewer 2: There isn't a single `Adapter` class. Instead there are several adaptor classes available for common Java types. For example the `HashSetAdapter`.

> * Add the `protostream-types` artifact to your Maven dependencies as follows:

Reviewer 2: All of the adapter classes provided by this jar are already registered by default in `core` and the `RemoteCacheManager`.

Maybe we should list of all these?

```
            // collections
            ArrayListAdapter.class,
            LinkedListAdapter.class,
            HashSetAdapter.class,
            LinkedHashSetAdapter.class,
            TreeSetAdapter.class,
            // arrays
            BooleanArrayAdapter.class,
            ShortArrayAdapter.class,
            IntArrayAdapter.class,
            LongArrayAdapter.class,
            FloatArrayAdapter.class,
            DoubleArrayAdapter.class,
            BoxedBooleanArrayAdapter.class,
            BoxedByteArrayAdapter.class,
            BoxedShortArrayAdapter.class,
            BoxedIntegerArrayAdapter.class,
            BoxedLongArrayAdapter.class,
            BoxedFloatArrayAdapter.class,
            BoxedDoubleArrayAdapter.class,
            StringArrayAdapter.class,
            ObjectArrayAdapter.class
            // common types
            UUIDAdapter.class,
            BigIntegerAdapter.class,
            BigDecimalAdapter.class
```

Reviewer 2: So I think we're talking about slightly different things here.

To store any of the ^ classes as a key/value directly the user doesn't have todo anything on the client or in embedded and no need to add the protostream-types dependency.

If the user wants to have one of ^ classes as a marshallable field in one of their Protostream annotated POJOS, then it's necessary for their `@AutoProtoSchemaBuilder` to use the `dependsOn`  field to depend on `org.infinispan.protostream.types.java.CommonTypesSchema` or `org.infinispan.protostream.types.java.CommonContainerTypesSchema`

Author: @ryanemerson I think we should list a couple common types to illustrate the point. when you import to an IDE it lists all the classes available so I don't think we should try and have a definitive enumeration of each class in the docs.

Reviewer 2: Yeah that's a fair point. Presuambly the list will expand over time as well.

I think it's important to document the requirement for setting `dependsOn` on the users `@AutoProtoSchemaBuilder` if they want to use any of these objects as fields in their Pojos.

Author: @ryanemerson OK, I've updated with this but I suppose there is still some discussion upstream but I think this comment is resolved.
